fix for Issue #8

J0F3 opened the pull request:
Pull Request (PR) description
Added a Check in the CheckForPreReqs Function so that the installation of the Windows Feature "RSAT-Feature-Tools-BitLocker-RemoteAdminTool" on Server Core because it is only available on Full Server.

This Pull Request (PR) fixes the following issues:
Fixes #8

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xbitlocker/9)
<!-- Reviewable:end -->

Commit:
fix for Server Core

Commit:
Added PowerShellVerion=4.0 added to Module Manifest

johlju commented:
@J0F3 If I review this PR do you have time to work on it?

johlju commented:
When you have resolved the review comments, or if you need to discuss a comment. Could you please go into Reviewable and write 'Done' (or click the done-button) on each of the review comments. This allows me to acknowledge (resolve) the review comments and I know that you are done with each change. After you written 'Done' on each review comment,  or comment on the review comment if it needs further discussion, then click the big green 'Publish' button on top of the page. That will send all your comments back as one big comment to this PR.
You find Reviewable if you click on the big purple Reviewable-button in the PR description.

Commit:
Update README according the review comments

Commit:
replaced double quotes with single quotes according the review comments

Commit:
Update code styling of Unit Test according the review comments

J0F3 commented:
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[README.md, line 8 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_oYnR15U4Ok7PHg0p:-LDvNKiN3sk8ksBsVQaa:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/f718d81689d87e51b5f2deb9d237bf378b7f57a4/README.md#L8)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should not remove this.
</blockquote></details>

Done.

---

*[README.md, line 102 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_ocvoDcSuBWMP7ZUD:-LDvNrtG9lICCproawep:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/README.md#L102)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

If an entry resolves an issue please reference the issue where applicable. You may suffix (it's optional) each entry with your name. Example of the format for the entry.
```
* Descriptive entry ([issue #123](https://github.com/PowerShell/xPSDesiredStateConfiguration/issues/123). [Name/Alias (@github_account)](https://github.com/github_account)`
```
</blockquote></details>

Done.

---

*[Misc/xBitlockerCommon.psm1, line 424 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_ogHZ4K7l-mgqfZwL:-LDvP-ME27_T-mjoW2p8:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Misc/xBitlockerCommon.psm1#L424)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> "Installed"
> ```

We should use single quotes where applicable. Throughout this row.
</blockquote></details>

Done.

---

*[Misc/xBitlockerCommon.psm1, line 523 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_ook7Clb0RX-HayU8:-LDvP5iV755-u4-ajD1M:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Misc/xBitlockerCommon.psm1#L523)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should use single quotes where applicable. 
</blockquote></details>

Done.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 5 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_p47N8nIlaDAd6MlA:-LDvPsPu9nz2daitUQb2:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L5)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This should move back one indent.
</blockquote></details>

Done.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 17 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pC7u4odxHJjCJ8Xl:-LDvQBy06bUk5Wlb7G9Z:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L17)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should use single quotes where applicable. Throughout this file.
</blockquote></details>

Done.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 17 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_p7DzCB-8gD_HJVC-:-LDvRxRjCNa1aoKMj4ix:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L17)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should start Context -block with 'When...'. Throughout.
</blockquote></details>

Done.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 33 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pKZi-Ixp53lO8A5F:-LDvQIYv6G3vQrdoXJHw:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L33)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should have open brace on a separate row. 
</blockquote></details>

Done.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 66 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pSXPCa6yP_ZRsO-0:-LDvQhzDC5Mq51Ju7jro:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L66)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This should use  `Should -Not -Throw' 
</blockquote></details>

Done.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 86 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pPpz8mnqT3r9a33m:-LDvRqjS2lh_R9OvL9ZI:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L86)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should have open brace on a separate row.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

Commit:
FIx erros in tests

J0F3 commented:
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 86 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pPpz8mnqT3r9a33m:-LDvWFNQEcpFQkubz9oL:b-dz7923) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L86)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can not do that because this will alter the logic of the mocked function. (return value will be null and not the hash table)

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 commented:
Sorry about the delay.   I have rebased the PR and looked through the comments now.
Thx

johlju commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 17 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pC7u4odxHJjCJ8Xl:-LE9snX97fQ12OeTfCHT:b-jm9rs6) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L17)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Please also use single quotes around description for Context- and It-blocks. Throughout.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 86 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pPpz8mnqT3r9a33m:-LE9tM4467pDP16iyiCj:b-sv8juh) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L86)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Can not do that because this will alter the logic of the mocked function. (return value will be null and not the hash table)
</blockquote></details>

Sorry, I'm not following you here :/ The code here looks good to me.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 23 at r3](https://reviewable.io/reviews/powershell/xbitlocker/9#-LE9u4fk4zulYZo7nHuM:-LE9u4fk4zulYZo7nHuN:b-6e4jwc) ([raw file](https://github.com/powershell/xbitlocker/blob/da084fd249b7eef45cff381a10d715b7f61b892a/Test/Unit/xBitlockerCommon.tests.ps1#L23)):*
> ```PowerShell
> 
>             Mock -CommandName Get-WindowsFeature -MockWith {
>                 param
> ```

We can move the param block to the 'empty function' so we don't have to duplicate that in every mock. Throughout.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 committed:
Add some code formating optimizations and merge file from 'Test' direcotory

J0F3 commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 23 at r3](https://reviewable.io/reviews/powershell/xbitlocker/9#-LE9u4fk4zulYZo7nHuM:-LEArBDL0cwDldpRDEAA:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/da084fd249b7eef45cff381a10d715b7f61b892a/Test/Unit/xBitlockerCommon.tests.ps1#L23)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We can move the param block to the 'empty function' so we don't have to duplicate that in every mock. Throughout.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 86 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pPpz8mnqT3r9a33m:-LEArohlDuD5N15ct6fe:b-diys05) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L86)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Sorry, I'm not following you here :/ The code here looks good to me.
</blockquote></details>

I thought that you mean the open brace of the hash table of the return value should be on a separate row but you probably meant the brace after the else just above. So everything fine anyway :-)

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 commented:
I also have just realized that we had now two unit tests for xBitlockerCommon.ps1. So I just merged mine into the file under Tests\Unit so we have just one file with all test in it.

johlju commented:
Looking good now! Just a tiny review comment left :slightly_smiling_face: 

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 193 at r4](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEE-ZWU3rGNC9g6RD9O:-LEE-ZWU3rGNC9g6RD9P:b-pe4nqi) ([raw file](https://github.com/powershell/xbitlocker/blob/1bac4b2e6d5bf2ee335748e40147e791a4f7ca11/Tests/Unit/xBitlockerCommon.tests.ps1#L193)):*
> ```PowerShell
>                     }
>                 }
>                 It 'The CheckForPreReqs function should throw an exceptions about missing Windows Feature' {
> ```

If we mock `Write-Error`we can get rid of the text that looks like an error in the tests output 
https://ci.appveyor.com/project/PowerShell/xbitlocker/build/1.1.54.0#L614

If we mock `Write-Error` then we could use this after the test for `Should -Throw` to verify that Write-Error was called correctly.

```
`Assert-MockCalled -Command `Write-Error` -Exactly -Time 1 -Scope It
```

*I throw me twice now thinking why the test fail but did not report failure* :smiley:

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 86 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pPpz8mnqT3r9a33m:-LEE0e6P3Zo5_8V1f7eJ:b1s2kur) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L86)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

I thought that you mean the open brace of the hash table of the return value should be on a separate row but you probably meant the brace after the else just above. So everything fine anyway :-)
</blockquote></details>

Ah, yes, it was the else open brace I meant :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9#-:-LEE0nrs7K9M819ibVfv:b-kxktlp)*
<!-- Sent from Reviewable.io -->

johlju commented:
There are merge conflicts since your PR was not based on the latest changes. Could you please rebase against branch dev using `git rebase` (not by using `git pull` or `git merge`, to keep the commit history). If you don't know how to rebase, please look at how to [Resolve merge conflicts](https://github.com/PowerShell/DscResources/blob/master/GettingStartedWithGitHub.md#resolve-merge-conflicts).
Let me know if you need any assistance.

J0F3 committed:
Add mock of Write-Error to get rid of errout output at line 203

J0F3 commented:
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 193 at r4](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEE-ZWU3rGNC9g6RD9O:-LEG9Rnz8Tbpy079sG-k:bblwg15) ([raw file](https://github.com/powershell/xbitlocker/blob/1bac4b2e6d5bf2ee335748e40147e791a4f7ca11/Tests/Unit/xBitlockerCommon.tests.ps1#L193)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

If we mock `Write-Error`we can get rid of the text that looks like an error in the tests output 
https://ci.appveyor.com/project/PowerShell/xbitlocker/build/1.1.54.0#L614

If we mock `Write-Error` then we could use this after the test for `Should -Throw` to verify that Write-Error was called correctly.

```
`Assert-MockCalled -Command `Write-Error` -Exactly -Time 1 -Scope It
```

*I throw me twice now thinking why the test fail but did not report failure* :smiley:
</blockquote></details>

Ah that's what I was looking into but could not find a solution how to get rid of the error output.
So lets try. Not sure if this is correct this way. Is it? At least the error output is now gone an the test look nice and green :-)

---

*[Test/Unit/xBitlockerCommon.tests.ps1, line 86 at r1](https://reviewable.io/reviews/powershell/xbitlocker/9#-LD_pPpz8mnqT3r9a33m:-LEG8zSN8GQSYJHJxP_0:b2oi3h1) ([raw file](https://github.com/powershell/xbitlocker/blob/c16071ba621f18e59790958d008ed4b465db26dc/Test/Unit/xBitlockerCommon.tests.ps1#L86)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Ah, yes, it was the else open brace I meant :)
</blockquote></details>


👍

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 193 at r4](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEE-ZWU3rGNC9g6RD9O:-LELiavIAkafJ_qIxaVo:b-5loop1) ([raw file](https://github.com/powershell/xbitlocker/blob/1bac4b2e6d5bf2ee335748e40147e791a4f7ca11/Tests/Unit/xBitlockerCommon.tests.ps1#L193)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Ah that's what I was looking into but could not find a solution how to get rid of the error output.
So lets try. Not sure if this is correct this way. Is it? At least the error output is now gone an the test look nice and green :-)
</blockquote></details>

Looking good! Just a few tweaks, and I think we are all good :slightly_smiling_face:  See other comments.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 33 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELgWoI1YF6fiuRrZfX:-LELgWoI1YF6fiuRrZfY:b-ppkipr) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L33)):*
> ```PowerShell
>         }
> 
>         function Write-Error
> ```

This should not be needed for cmdlets that are part of "default PowerShell". LEt's try without this.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 199 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELgnstAdeOxrVboyp0:-LELgnstAdeOxrVboyp1:bl3rf0c) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L199)):*
> ```PowerShell
>                 }
>                 It 'The CheckForPreReqs function should throw an exceptions about missing Windows Feature' {
>                     Mock -CommandName Write-Error -MockWith {
> ```

I think it should be enough with `Mock -CommandName Write-Error`. We should not need `-MockWith` on this, because we want to test so that `throw` statement work on line 437 in the code?
If we just do an empty mock of  `Write-Error` then we can test so it gets called with the `Assert-MockCalled` you added.

If `Write-Error` gets called more than once, change `-Times 1` to `-Times 2` in the `Assert-MockCalled`etc.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 204 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELiSF546uGcYDJxlhB:-LELiSF546uGcYDJxlhC:bmanxpm) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L204)):*
> ```
> -Time
> ```

Should be `-Times` (ending with 's'). I spelled it wrong in the previous comment. :/

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 204 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELjfFs9CtlccyDWuOY:-LELjfFs9CtlccyDWuOZ:bq3qtnv) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L204)):*
> ```PowerShell
> 
>                     {CheckForPreReqs} | Should -Throw 'Required Bitlocker features need to be installed before xBitlocker can be used'
>                     Assert-MockCalled -Command Write-Error -Exactly -Time 1 -Scope It
> ```

This comment is non-blocking, meaning code is good as-is, but mentioning if you want to try it.
If you do want to test if `Write-Error` is actually called with the right message, then you can add this to the `Assert-MockCalled` (assuming I got the correct error message :) 

```
Assert-MockCalled -Command Write-Error -Exactly -Times 1 -Scope It -ParameterFilter {
    $Message -eq 'The RSAT-Feature-Tools-BitLocker-RemoteAdminTool feature needs to be installed before the xBitlocker module can be used'
}
```


---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 33 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELgWoI1YF6fiuRrZfX:-LEczdHA35I8kZM6JLRZ:b6e36lp) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L33)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This should not be needed for cmdlets that are part of "default PowerShell". LEt's try without this.
</blockquote></details>

Yup, works without. :+1: 

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 199 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELgnstAdeOxrVboyp0:-LEd5Nfm3r9WDaxgPXno:b-i3i4xw) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L199)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

I think it should be enough with `Mock -CommandName Write-Error`. We should not need `-MockWith` on this, because we want to test so that `throw` statement work on line 437 in the code?
If we just do an empty mock of  `Write-Error` then we can test so it gets called with the `Assert-MockCalled` you added.

If `Write-Error` gets called more than once, change `-Times 1` to `-Times 2` in the `Assert-MockCalled`etc.


</blockquote></details>

Ah, learned again something. thx :-) (removed the -MockWith)

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 204 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELiSF546uGcYDJxlhB:-LEd5mkoAPfz4TieJQVj:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L204)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Time
> ```

Should be `-Times` (ending with 's'). I spelled it wrong in the previous comment. :/
</blockquote></details>

Done.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 204 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELjfFs9CtlccyDWuOY:-LEd5ndAD8uC9zcah_DD:bcwmjp2) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L204)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This comment is non-blocking, meaning code is good as-is, but mentioning if you want to try it.
If you do want to test if `Write-Error` is actually called with the right message, then you can add this to the `Assert-MockCalled` (assuming I got the correct error message :) 

```
Assert-MockCalled -Command Write-Error -Exactly -Times 1 -Scope It -ParameterFilter {
    $Message -eq 'The RSAT-Feature-Tools-BitLocker-RemoteAdminTool feature needs to be installed before the xBitlocker module can be used'
}

</blockquote></details>

I am stuck here. Can not get to work this. What should be Line 195 be then? Still `Should -Throw 'Required Bitlocker features need to be installed before xBitlocker can be used'` or only `Should -Throw`?

Pester says now `Expected Write-Error in module xBitlockerCommon to be called 1 times exactly but was called 0 times`

But when I change it to the this: 
`                    
Assert-MockCalled -Command Write-Error -Exactly -Times 1 -Scope It -ParameterFilter {
                        $Message
                        $Message -eq 'Required Bitlocker features need to be installed before xBitlocker can be used'
                    }
`
Makes not really sense to me. What I am missing here?

Thanks for you help! :-)

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

Commit:
Add test case for server core without required featues and tests for error messages

johlju commented:
Review status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @johlju)

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 204 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELjfFs9CtlccyDWuOY:-LEi46jp3bg7uRQRlKDJ:b-hyu05g) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L204)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

I am stuck here. Can not get to work this. What should be Line 195 be then? Still `Should -Throw 'Required Bitlocker features need to be installed before xBitlocker can be used'` or only `Should -Throw`?

Pester says now `Expected Write-Error in module xBitlockerCommon to be called 1 times exactly but was called 0 times`

But when I change it to the this: 
`                    
Assert-MockCalled -Command Write-Error -Exactly -Times 1 -Scope It -ParameterFilter {
                        $Message
                        $Message -eq 'Required Bitlocker features need to be installed before xBitlocker can be used'
                    }
`
Makes not really sense to me. What I am missing here?

Thanks for you help! :-)
</blockquote></details>

Happy to help. It should look something like this. `throw` is called from the code, so `| Should -Throw` will catch that, we should not mock that. We want to test so the code throws, not our mock. 

NOTE! You must remove the stub `function Write-Error` you have in the beginning of the InModuleScope-block for this to work (see a review comment).
The `Mock` command will mock all the parameters from the original command, then we can use `ParameterFilter`. When there are no parameters in the stub function the filter won't work. Also we should not make stub functions for cmdlets that are native to PowerShell (native as exist by default without needing to install a module). :) 

```
It 'The CheckForPreReqs function should throw an exceptions about missing Windows Feature' {
    Mock -CommandName Write-Error

    {CheckForPreReqs} | Should -Throw 'Required Bitlocker features need to be installed before xBitlocker can be used'
    Assert-MockCalled -Command Write-Error -ParameterFilter {
        $Message -match 'The RSAT-Feature-Tools-BitLocker-RemoteAdminTool feature needs to be installed before the xBitlocker module can be used'
    } -Exactly -Times 1 -Scope It
}
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

Commit:
Add tests for Get-OSEdition

J0F3 commented:
Review status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @johlju)

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 204 at r6](https://reviewable.io/reviews/powershell/xbitlocker/9#-LELjfFs9CtlccyDWuOY:-LEiA_xF4svEyvYat07e:b64zwsd) ([raw file](https://github.com/powershell/xbitlocker/blob/a44d014bbdb7d1623b1cca8ba822cdcb5c09f26e/Tests/Unit/xBitlockerCommon.tests.ps1#L204)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Happy to help. It should look something like this. `throw` is called from the code, so `| Should -Throw` will catch that, we should not mock that. We want to test so the code throws, not our mock. 

NOTE! You must remove the stub `function Write-Error` you have in the beginning of the InModuleScope-block for this to work (see a review comment).
The `Mock` command will mock all the parameters from the original command, then we can use `ParameterFilter`. When there are no parameters in the stub function the filter won't work. Also we should not make stub functions for cmdlets that are native to PowerShell (native as exist by default without needing to install a module). :) 

```
It 'The CheckForPreReqs function should throw an exceptions about missing Windows Feature' {
    Mock -CommandName Write-Error

    {CheckForPreReqs} | Should -Throw 'Required Bitlocker features need to be installed before xBitlocker can be used'
    Assert-MockCalled -Command Write-Error -ParameterFilter {
        $Message -match 'The RSAT-Feature-Tools-BitLocker-RemoteAdminTool feature needs to be installed before the xBitlocker module can be used'
    } -Exactly -Times 1 -Scope It
}
```
</blockquote></details>

Thank you very much! I got it now (I think :-) ). The error output was actually from the Write-Error statements in the code and not from the Throw. I miss interpreted that. So I rewritten the whole this a little bit and added also a test case for server core without the needed features which was completely missing before. Further I added also some test for the Get-OSEdition function so we have that also covered.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 197 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEiBc921uKq2ZIPoFy0:-LEiBc933rg_FHMDM08E:b-n2fo0c) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L197)):*
> ```PowerShell
> 
>                 It 'Should give an error that Bitlocker Windows Feature needs to be installed' {
>                     {CheckForPreReqs} | Should -Throw
> ```

Not sure if this is the right solution. Actually I testing here always two things. But could not find any other way to suppress the exception from throw so I can only test for the Write-Error messages.
And, I think, all together in one It statement would also not be ideal.
So open for suggestion if there is a better way to test only the Write-Error messages while ignoring the throw exception (which get tested separate).
Thx!

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed, 3 unresolved discussions (waiting on @J0F3)

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 141 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEiu1mBCTpXfncrpaA3:-LEiu1mBCTpXfncrpaA4:bpvjaev) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L141)):*
> ```
> {}
> ```

We don't need this when we don't mock any logic. This can be removed. Throughout the mocks.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 143 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEitpqX35zw7kyDO1PU:-LEitpqX35zw7kyDO1PV:b-5p935n) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L143)):*
> ```
> -Time 0
> ```

Should be `-Times 0`. Please change to `-Times` (with 's') throughout the tests.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 197 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEiBc921uKq2ZIPoFy0:-LEiubPD0CCqwV3clQ51:bnt1cjn) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L197)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Not sure if this is the right solution. Actually I testing here always two things. But could not find any other way to suppress the exception from throw so I can only test for the Write-Error messages.
And, I think, all together in one It statement would also not be ideal.
So open for suggestion if there is a better way to test only the Write-Error messages while ignoring the throw exception (which get tested separate).
Thx!
</blockquote></details>

I don't think there is a way to mock `throw`, so separating them is not possible. 

I think this is the correct way. You want to test what happens in the code, and the code calls `Write-Error` and then throws. So it's correct to test that in the same It-block. 

*The problem is probably the code. It should just build a string and throw one correct error message, or use `Write-Warning` instead of `Write-Error`. Using `Write-Error` several times and then throw seems wrong. But that can be another PR.* :slightly_smiling_face: 

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 commented:
Review status: all files reviewed, 2 unresolved discussions (waiting on @johlju)

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 141 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEiu1mBCTpXfncrpaA3:-LEkXLyR6UWISUwnWNNo:b-896fix) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L141)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> {}
> ```

We don't need this when we don't mock any logic. This can be removed. Throughout the mocks.
</blockquote></details>

Done.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 143 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEitpqX35zw7kyDO1PU:-LEkXN4z7Q0rYTt090Rw:b37refa) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L143)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Time 0
> ```

Should be `-Times 0`. Please change to `-Times` (with 's') throughout the tests.
</blockquote></details>

Doh! Missed that again... 🤦‍♂️
But interesting that it work also without the 's'.

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 197 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEiBc921uKq2ZIPoFy0:-LEkYC8NB0j-7wgomu7y:b45idog) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L197)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

I don't think there is a way to mock `throw`, so separating them is not possible. 

I think this is the correct way. You want to test what happens in the code, and the code calls `Write-Error` and then throws. So it's correct to test that in the same It-block. 

*The problem is probably the code. It should just build a string and throw one correct error message, or use `Write-Warning` instead of `Write-Error`. Using `Write-Error` several times and then throw seems wrong. But that can be another PR.* :slightly_smiling_face: 
</blockquote></details>

Ok, fine. 
Yes I thought the same. The code is a little bit strange. However it was already there an I only added the adjustment for the missing Windows Features on Core Server.
Maybe it would also make more sense when the function just returns $true or $false and the exception is throw in the code which calls the function. But as you wrote that would be an other PR. :-)

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9)*
<!-- Sent from Reviewable.io -->

J0F3 committed:
Remove curly brackets for Write-Error moks and replace -Time with -Times for Assert-MockCalled

johlju commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r8.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 143 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEitpqX35zw7kyDO1PU:-LEmnRneAvujypLeOTFy:bsibot) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L143)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Doh! Missed that again... 🤦‍♂️
But interesting that it work also without the 's'.
</blockquote></details>

You only have to specify the name of a named parameter until it's unique, like `Write-Verbose -M 'hej' -V` works. But for the style guideline we have to write out the entire named parameter. :slightly_smiling_face:  

---

*[Tests/Unit/xBitlockerCommon.tests.ps1, line 197 at r7](https://reviewable.io/reviews/powershell/xbitlocker/9#-LEiBc921uKq2ZIPoFy0:-LEmoE0f195zJBPtQuSI:bfra1fw) ([raw file](https://github.com/powershell/xbitlocker/blob/5addd56cc488307a21bca3cb2014d13df1d25eae/Tests/Unit/xBitlockerCommon.tests.ps1#L197)):*
<details><summary><i>Previously, J0F3 (Jonas Feller) wrote…</i></summary><blockquote>

Ok, fine. 
Yes I thought the same. The code is a little bit strange. However it was already there an I only added the adjustment for the missing Windows Features on Core Server.
Maybe it would also make more sense when the function just returns $true or $false and the exception is throw in the code which calls the function. But as you wrote that would be an other PR. :-)
</blockquote></details>

That solution returning a Boolean would be even better yes. :slightly_smiling_face:  Good as-is for now. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xbitlocker/9#-:-LEmogSs2gVo2METcr-n:bnfp4nl)*
<!-- Sent from Reviewable.io -->

johlju commented:
@J0F3 Awesome work on this one! Thanks! 😃

J0F3 commented:
Thanks @johlju and thank you for your support! 😊
